### DISCUSSION TITLE
SIGBUS when loading arpack on Apple Silicon

### DISCUSSION THREAD

**Discussion by phst** [Homebrew/discussions#415] [Everyday usage] [answered]
Example:

```
$ python3.9 -m scipy.stats
Bus error: 10
```

This happens with an up-to-date Homebrew and Python + Numpy + Scipy installed from Homebrew on Apple Silicon. Here's an excerpt from the crash log:

```
Exception Type:        EXC_BAD_ACCESS (SIGBUS)
Exception Codes:       EXC_ARM_DA_ALIGN at 0xf9000100d503201f
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Bus error: 10
Termination Reason:    Namespace SIGNAL, Code 0xa
Terminating Process:   exc handler [73194]

VM Regions Near 0x100d503201f:
--> GPU Carveout (reserved)    1000000000-7000000000   [384.0G] ---/--- SM=NUL  reserved VM address space (unallocated)
    

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   ???                           	0x00000100d503201f 0 + 1103085379615
1   _arpack.cpython-39-darwin.so  	0x0000000116cded54 PyFortranObject_New + 40
2   _arpack.cpython-39-darwin.so  	0x0000000116cd6ba0 PyInit__arpack + 1024
3   org.python.python             	0x0000000100db1378 _PyImport_LoadDynamicModuleWithSpec + 668
4   org.python.python             	0x0000000100db0be4 _imp_create_dynamic + 340
5   org.python.python             	0x0000000100d1c864 cfunction_vectorcall_FASTCALL + 248
6   org.python.python             	0x0000000100d92668 _PyEval_EvalFrameDefault + 24716
7   org.python.python             	0x0000000100d953a0 _PyEval_EvalCode + 1932
8   org.python.python             	0x0000000100ce8ad8 _PyFunction_Vectorcall + 244
9   org.python.python             	0x0000000100d948a8 call_function + 416
```

So this looks like `arpack` as used by Scipy doesn't correctly align data memory for ARM64.
Can others reproduce this?

**Comment by CharlesSOo** [comment]
Can't seem to put my finger on this, but this seems to be happening a lot for recent users not just in the homebrew space. 

Are you seeing high usage from `mdworker_shared`

**Reply by phst** [reply]
> Can't seem to put my finger on this, but this seems to be happening a lot for recent users not just in the homebrew space.

I think this is typically a result of incorrectly-aligned loads and stores, e.g.

```c
char *p = ...
int i = *(int *)(p + 1);
```

These tend to work on x86, but ARM normally requires aligned memory accesses, and I don't know whether the compilers are smart enough to use the unaligned instructions in cases like this.

> Are you seeing high usage from `mdworker_shared`

How would that be related? `mdworker_shared` is the Spotlight indexer and shouldn't affect generation of invalid ARM opcodes.

**Comment by xssfox** [comment]
@phst did you happen to find a solution to this? I've just come across this issue myself (though with `from scipy.signal import butter, lfilter`)

**Reply by phst** [reply]
No, I was mainly asking around whether it's reproducible on other machines, following the suggestions at https://github.com/Homebrew/homebrew-core/issues/new/choose. Since that seems to be the case, I'll report a bug.

**Reply by phst** [reply]
Bug report is at https://github.com/Homebrew/brew/issues/10343.

**Reply by carlocab** [reply]
That looks like a formula-related bug, and not a bug with `brew` itself, so the report should've been at https://github.com/Homebrew/homebrew-core. Someone will probably transfer your issue later, maybe.

**Comment by xght** [comment]
Same issue for me.
The problem seems to be in _f2pyinittiming_ : it jumps to 0x00000100d503201f which isn't mapped.
The funny thing is that 0xD503201F (the low part of the address) corresponds to the nop in ARM64.
I don't know if I will find time to debug this further.

**Reply by xght** [reply]
Generated code for _f2pyinittiming_ decompiled in ghidra (I don't know where this function comes from) :

```c
void _f2pyinittiming_(code *param_1)
{
  if (((uint)param_1 >> 2 & 1) != 0) {
    param_1 = *(code **)(param_1 + 4);
  }
  (*param_1)(&_timing_,0x76984,&DAT_00076988,0x7698c,&DAT_00076990,&DAT_00076994,&DAT_00076998,
             0x7699c,&DAT_000769a0,0x769a4,&DAT_000769a8,&DAT_000769ac,&DAT_000769b0,&DAT_000769b4,
             &DAT_000769b8,&DAT_000769bc,&DAT_000769c0,&DAT_000769c4,&DAT_000769c8,&DAT_000769cc,
             &DAT_000769d0,&DAT_000769d4,&DAT_000769d8,&DAT_000769dc,&DAT_000769e0,&DAT_000769e4,
             &DAT_000769e8,&DAT_000769ec,&DAT_000769f0,&DAT_000769f4,&DAT_000769f8);
  return;
}
```

param_1 is a function pointer but is 4 bytes aligned, not 8 bytes aligned, so the if branch is taken and that leads to the crash.

I manually patched the _arpack.cpython-39-darwin.so and 6 others .so with similar patterns and I got a working version.

The problem may be in gfortran or in f2py.

**Reply by phst** [reply]
Thanks for the investigation! It looks like these functions are generated by f2py (e.g. search for `f2pyinit` in https://github.com/numpy/numpy/blob/c2b9d20cd6a8d28785972bd883f11d1a7fc3a8fa/numpy/f2py/common_rules.py). But I don't see anything wrong there – the function pointer argument passed to `f2pyinittiming` is `f2py_setup_timing`, and that's a normal static function which should be aligned properly.
(But this is just a guess, I have no experience with either ARM64 or Fortran.)

**Reply by xght** [reply]
```asm
_f2py_init_timing                               XREF[1]:     _PyInit__arpack:00002b94(*)  
00002be0 20 02 04 10     adr        x0=>_f2py_setup_timing,0xac24
00002be4 1f 20 03 d5     nop
00002be8 fa 27 00 14     b          _f2pyinittiming_                                 undefined _f2pyinittiming_()
```

You can see that the address of `_f2py_setup_timing` (base offset + 0xac24) is not 8 aligned as expected by _f2pyinittiming_. 
(A 4 bytes alignement is completely normal for a function in ARM64)

I found the autogenerated code for  `_f2py_setup_timing` in a file named `_arpack-f2pywrappers.f`.
```fortran
subroutine f2pyinittiming(setupfunc)
      external setupfunc
      integer nopx
      integer nbx
      integer nrorth
      integer nitref
      integer nrstrt
      real tsaupd
      real tsaup2
      real tsaitr
      real tseigt
      real tsgets
      real tsapps
      real tsconv
      real tnaupd
      real tnaup2
      real tnaitr
      real tneigh
      real tngets
      real tnapps
      real tnconv
      real tcaupd
      real tcaup2
      real tcaitr
      real tceigh
      real tcgets
      real tcapps
      real tcconv
      real tmvopx
      real tmvbx
      real tgetv0
      real titref
      real trvec
      common /timing/ nopx,nbx,nrorth,nitref,nrstrt,tsaupd,tsaup2,
     &tsaitr,tseigt,tsgets,tsapps,tsconv,tnaupd,tnaup2,tnaitr,tneigh,tng
     &ets,tnapps,tnconv,tcaupd,tcaup2,tcaitr,tceigh,tcgets,tcapps,tcconv
     &,tmvopx,tmvbx,tgetv0,titref,trvec
      call setupfunc(nopx,nbx,nrorth,nitref,nrstrt,tsaupd,tsaup2,t
     &saitr,tseigt,tsgets,tsapps,tsconv,tnaupd,tnaup2,tnaitr,tneigh,tnge
     &ts,tnapps,tnconv,tcaupd,tcaup2,tcaitr,tceigh,tcgets,tcapps,tcconv,
     &tmvopx,tmvbx,tgetv0,titref,trvec)
      end
```

Nothing wrong here neither.

**Reply by xght** [reply]
The culprit is gfortran : 
 
```fortran
        subroutine f2pyinittiming(setupfunc)
                external setupfunc
                call setupfunc()
        end
```
in arm64 with `gfortran --shared -O2` it gives:

```asm
0000000000003fb0 _f2pyinittiming_:
    3fb0: 60 00 10 37                  	tbnz	w0, #2, #12 <_f2pyinittiming_+0xc>
    3fb4: f1 03 00 aa                  	mov	x17, x0
    3fb8: 20 02 1f d6                  	br	x17
    3fbc: 10 c0 5f f8                  	ldur	x16, [x0, #-4]
    3fc0: 00 40 40 f8                  	ldur	x0, [x0, #4]
    3fc4: fc ff ff 17                  	b	#-16 <_f2pyinittiming_+0x4>
```

and in amd64: 

```asm
0000000000001100 <f2pyinittiming_>:                                                                                                                    
    1100:       31 c0                   xor    %eax,%eax                                                                                               
    1102:       ff e7                   jmpq   *%rdi
```

There is this weird alignement check (the tbnz) that causes the SIGBUS in ARM64 but it is not present in x64.

**Reply by hjmr** [reply]
> I manually patched the _arpack.cpython-39-darwin.so and 6 others .so with similar patterns and I got a working version.
> The problem may be in gfortran or in f2py.

Could you tell me how to manually patch .so files? Thanks in advance.

**Reply by xght** [reply]
It will require a little bit of assembly knowledge, and you muss resign the code (I struggled a lot with that).

You need to find the code that crashes and nop the tbnz (nop is 0xD503201F in ARM64).

Patch the first library (_arpack.cpython-39-darwin.so), resign it, and you will see that there is another crash in another library, it's the same problem each time. 

I will try to make a diff.

**Reply by hjmr** [reply]
@xght Thank you for reply. I tried replacing codes with NOP where it crashed at, and resign the library but not worked well. I don't know if I could replace the correct place with the correct codes :-(  I will try some more. Anyway, thank you for the hint!

**Reply by xght** [reply]
For arpack the patch of the `_f2pyinittiming_` is:
`at offset 0x0000cc80 replace a8 01 10 37 by 1f 20 03 d5`

Good luck for the others .so, sometimes you will need to patch the function called `_f2pyinittypes_`.

Or wait for the fix, it will be cleaner !

**Reply by hjmr** [reply]
@xght Thank you very much. I have patched several .so files using hexedit and can finally get the scipy working again. The Online Disassembler was the great help to do this. But this is quite exhausting work and I feel it's better to wait for the fix. Anyway, thank you!

**Comment by xght** [comment, marked as answer]
The bug is already reported at https://github.com/iains/gcc-darwin-arm64/issues/32

**Reply by phst** [reply]
Thanks! For reference, https://github.com/scipy/scipy/issues/13364 is a similar bug report directly for Scipy.

**Reply by phst** [reply]
Marking this answer as accepted since the fix can be discussed on those bug reports.
